doc/glossary.rst

```diff
@@ -0,0 +1,973 @@
+        (Exceptions, for legacy reasons, include
+        :class:`~pipeline.Pipeline` and
+        :class:`~pipeline.FeatureUnion`.)
+
+    common tests
+        This refers to the tests run on almost every estimator class in
+        Scikit-learn to check they comply with basic API conventions.  They are
+        available for external use through
+        :func:`utils.estimator_checks.check_estimator`.
+
+        Note: Some exceptions to the common testing regime are currently
+        hard-coded into the library, but we hope to replace this by marking
+        exceptional behaviours on the estimator using semantic :term:`estimator
+        tags`.
+
+    deprecation
+        We use deprecation to slowly violate our :term:`backwards
+        compatibility` assurances, usually to to:
+
+        * change the semantics of the default value of a parameter; or
+        * remove a parameter, attribute, method, class, etc.
+
+        We will ordinarily issue a warning when a deprecated element is used,
+        although there may be limitations to this.  For instance, we will raise
+        a warning when someone sets a parameter that has been deprecated, but
+        may not when they access that parameter's attribute on the estimator
+        instance.
+
+        See the :ref:`Contributors' Guide <contributing_deprecation>`.
+
+    docstring
+        TODO
+
+    double underscore notation
+        When specifying parameter names for nested estimators, ``__`` may be
+        used to separate between parent and child.
+        See :term:`parameter`.
+
+    dtype
+    data type
+        TODO. Mention casting.
+
+    duck typing
+        TODO
+        Note that ``getattr`` should be preferred to ``hasattr`` since
+        ``hasattr`` can be expensive, particularly for some model attributes.
+
+    estimator instance
+        We sometimes use this terminology to distinguish an :term:`estimator`
+        class from a constructed instance. For example, in the following,
+        ``cls`` is an estimator class, while ``est1`` and ``est2`` are
+        instances::
+
+            cls = RandomForestClassifier
+            est1 = cls()
+            est2 = RandomForestClassifier()
+
+    examples
+        We try to give examples of basic usage for most functions and
+        classes in the API:
+
+        * as doctests in their docstrings (i.e. within the ``sklearn/`` library
+          code itself).
+        * as examples in the :ref:`example gallery <general_examples>`
+          rendered from scripts in the ``examples/`` directory, exemplifying
+          key features or parameters of the estimator/function.
+        * sometimes in the :ref:`User Guide <user_guide>` (built from ``doc/``)
+          alongside a technical description of the estimator.
+
+    evaluation metric
+        TODO
+
+    estimator tags
+        A proposed feature (e.g. :issue:`8022`) by which the capabilities of an
```

Comment:
It's interesting that you deem this worthy of an entry even though it's not merged yet ;) Not that I disagree. I just gave it to a student.

Reply:
I hope that this document is able to capture some of the tacit knowledge held by core contributors. I think occasional mention to long-roadmapped features doesn't hurt and might help. The main risk is keeping this up-to-date.